File: cpp_src/CombinationSum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        int size = candidates.size();
        if(size == 0 || target == 0){
            return result;
        }
        sort(candidates.begin(),candidates.end());
        vector<int> oneCase;
        recursive(candidates,0,target,oneCase,result);
        return result;
    }
    
    void recursive(vector<int>& candidates, int start, int target, vector<int>& oneCase, vector<vector<int>>& result){
        int size = candidates.size();
        for(int i = start; i < size; i++){
            target -= candidates[i];
            if(target < 0){
                return;
            }else if(target == 0){
                oneCase.push_back(candidates[i]);
                vector<int> final(oneCase);
                result.push_back(final);
                oneCase.pop_back();
                return;
            }else{
                oneCase.push_back(candidates[i]);
                recursive(candidates,i,target,oneCase,result);
            }
            target += candidates[i];
            oneCase.pop_back();
        }
    }
};

File: python_src/CombinationSum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        result = self.recursive(candidates, 0, target)
        for item in result:
            item.reverse()
        return result
    
    def recursive(self, candidates, index, target):
        if candidates[index] == target:
            return [[candidates[index]]]
        elif candidates[index] > target:
            return []
        else:
            result = []
            for i in range(index, len(candidates)):
                if target - candidates[i] == 0: 
                    result += [[candidates[i]]]
                    break
                ret = self.recursive(candidates, i, target - candidates[i])
            
                for item in ret:
                    item.append(candidates[i])
                result += ret
            return result
        